src/bin/drag_and_drop_listbox_model.rs

```diff
@@ -0,0 +1,653 @@
+        // caller, or store whatever new value the caller is providing.
+        //
+        // This maps between the GObject properties and our internal storage of the
+        // corresponding values of the properties.
+        impl ObjectImpl for RowData {
+            glib_object_impl!();
+
+            fn set_property(&self, _obj: &glib::Object, id: usize, value: &glib::Value) {
+                let prop = &PROPERTIES[id];
+
+                match *prop {
+                    subclass::Property("name", ..) => {
+                        let name = value.get();
+                        self.name.replace(name);
+                    }
+                    subclass::Property("count", ..) => {
+                        let count = value.get().expect("Got value of wrong type");
+                        self.count.replace(count);
+                    }
+                    _ => unimplemented!(),
+                }
+            }
+
+            fn get_property(&self, _obj: &glib::Object, id: usize) -> Result<glib::Value, ()> {
+                let prop = &PROPERTIES[id];
+
+                match *prop {
+                    subclass::Property("name", ..) => Ok(self.name.borrow().to_value()),
+                    subclass::Property("count", ..) => Ok(self.count.borrow().to_value()),
+                    _ => unimplemented!(),
+                }
+            }
+        }
+    }
+
+    // Public part of the RowData type. This behaves like a normal gtk-rs-style GObject
+    // binding
+    glib_wrapper! {
+        pub struct RowData(Object<subclass::simple::InstanceStruct<imp::RowData>, subclass::simple::ClassStruct<imp::RowData>, RowDataClass>);
+
+        match fn {
+            get_type => || imp::RowData::get_type().to_glib(),
+        }
+    }
+
+    // Constructor for new instances. This simply calls glib::Object::new() with
+    // initial values for our two properties and then returns the new instance
+    impl RowData {
+        pub fn new(name: &str, count: u32) -> RowData {
+            glib::Object::new(Self::static_type(), &[("name", &name), ("count", &count)])
+                .expect("Failed to create row data")
+                .downcast()
+                .expect("Created row data is of wrong type")
+        }
+    }
+}
+
+// return the previous visible row
+fn get_previous_row(list: &gtk::ListBox, row: &gtk::ListBoxRow) -> Option<gtk::ListBoxRow> {
+    if let Some(w) = list.get_row_at_index(row.get_index() - 1) {
+        if w.get_child()
+            .and_then(|w| w.downcast::<gtk::Revealer>().ok())
+            .map_or(false, |r| !r.get_reveal_child())
+        {
+            get_previous_row(list, &w)
+        } else {
+            Some(w)
+        }
+    } else {
+        None
+    }
+}
+
+// return the next visible row
+fn get_next_row(list: &gtk::ListBox, row: &gtk::ListBoxRow) -> Option<gtk::ListBoxRow> {
+    if let Some(w) = list.get_row_at_index(row.get_index() + 1) {
```

Comment:
What if `row.get_index() + 1 >= num_rows`?

Reply:
It returns null if there is no row on the specified index. 
https://gtk-rs.org/docs/gtk/trait.ListBoxExt.html#tymethod.get_row_at_index